File: interfaces/Matlab/example/example_hs015.m
%% HS015 problem
clc, clear

%% Optimization model
% Problem type: 'L' = Linear, 'Q' = Quadratic, 'N' = Nonlinear
model.problem_type = 'N';
% Vector indexing
model.base_indexing = 1;
% Number of variables
model.number_variables = 2;
% Variable bounds
model.variables_lower_bounds = [-inf, -inf]; 
model.variables_upper_bounds = [0.5, inf];

% Optimization sense: 1 = Minimize, -1 = Maximize
model.optimization_sense = 1;
% Objective function handle: objective = objective_function(x)
model.objective_function = @objective_function; 
% Objective gradient handle: gradient = objective_gradient(x)
model.objective_gradient = @objective_gradient;

% Number of constraints
model.number_constraints = 2;
% Constraint bounds
model.constraints_lower_bounds = [1.0, 0.0];
model.constraints_upper_bounds = [inf, inf];
% Constraint function handle: constraints = constrain_function(x)
model.constraint_function = @constraint_function;
% Constraint jacobian handle: jacobian = constraint_jacobian(x)
model.constraint_jacobian = @constraint_jacobian;
% Constraint sparsity pattern (base_indexing-based)
model.number_jacobian_nonzeros = 4;
model.jacobian_row_indices = [1 2 1 2];
model.jacobian_column_indices = [1 1 2 2];

% Lagrangian sign convention: 1 = rho*f(x) + y^T c(x), -1 = rho*f(x) - y^T c(x)
model.lagrangian_sign_convention = -1;
% Hessian triangular part: 'L' = lower, 'U' = upper
model.hessian_triangular_part = 'L';
% Lagrangian Hessian handle: hessian = lagrangian_hessian(x,rho,y)
model.lagrangian_hessian = @lagrangian_hessian;
% Lagrangian Hessian sparsity pattern (base_indexing-based)
model.number_hessian_nonzeros = 3;
model.hessian_row_indices = [1 2 2];
model.hessian_column_indices = [1 1 2];

% Jacobian operator handle: result = jacobian_operator(x,v)
% model.jacobian_operator = @jacobian_operator;

% Jacobian tranposed operator handle: result = jacobian_transposed_operator(x,v)
% model.jacobian_transposed_operator = @jacobian_transposed_operator;

% Hessian operator handle: result = lagrangian_hessian_operator(x,rho,y,v)
% model.lagrangian_hessian_operator = @lagrangian_hessian_operator;

% Initial primal point
model.initial_primal_iterate = [-2.0, 1.0];
% Dual primal point
model.initial_dual_iterate = [0, 0];

%% UNO solver options
% Preset can be possibly given: options = uno_options([preset])
options = uno_options(); % default preset
% options = uno_options('ipopt');

%% UNO solver callbacks
callbacks = struct();
% Logger stream handle: logger_stream_callback(str)
% callbacks.logger_stream_callback = @logger_stream_callback;
% Notify acceptable iterate handle: notify_acceptable_iterate_callback(x, yl, yb, y, rho, feas, stat, compl)
% callbacks.notify_acceptable_iterate_callback = @notify_acceptable_iterate_callback;
% Notify new primal handle: notify_new_primals(x)
% callbacks.notify_new_primals_callback = @notify_new_primals_callback;
% Notify new dual handle: notify_new_multipliers(x, yl, yb, y)
% callbacks.notify_new_multipliers_callback = @notify_new_multipliers_callback;
% User termination handle: terminate = user_termination_callback(x, yl, yb, y, rho, feas, stat, compl)
% callbacks.user_termination_callback = @user_termination_callback;

%% Solving the model
result = uno_optimize(model, options, callbacks);
% Clear mex functions
clear uno_optimize uno_options

%% Display the result
disp(result)

%% Model functions
% Objective function
function fval = objective_function(x)
    fval = 100 * (x(2) - x(1)^2)^2 + (1 - x(1))^2;
end

% Constraint function
function c = constraint_function(x)
    c = zeros(2,1);
    c(1) = x(1)*x(2);
    c(2) = x(1) + x(2)^2;
end

% Objective gradient
function grad = objective_gradient(x)
    grad = zeros(2,1);
    grad(1) = 400*x(1)^3 - 400*x(1)*x(2) + 2*x(1) - 2;
    grad(2) = 200*(x(2) - x(1)^2);
end

% Constraint jacobian
function J = constraint_jacobian(x)
    J = zeros(4,1);
    J(1) = x(2);
    J(2) = 1;
    J(3) = x(1);
    J(4) = 2*x(2);
end

% Lagrangian Hessian - lower triangular part
function H = lagrangian_hessian(x, objective_multiplier, multipliers)
    H = zeros(3,1);
    H(1) = objective_multiplier*(1200*x(1)^2 - 400*x(2) + 2);
    H(2) = -400*objective_multiplier*x(1) - multipliers(1);
    H(3) = 200*objective_multiplier - 2*multipliers(2);
end

% Lagrangian Hessian operator
function result = lagrangian_hessian_operator(x, objective_multiplier, multipliers, vector)
    h00 = objective_multiplier*(1200*x(1)^2 - 400*x(2) + 2);
    h10 = -400*objective_multiplier*x(1) - multipliers(1);
    h11 = 200*objective_multiplier - 2*multipliers(2);

    result = zeros(2,1);
    result(1) = h00*vector(1) + h10*vector(2);
    result(2) = h10*vector(1) + h11*vector(2);
end

File: interfaces/Python/example/example_hs015.py
# Copyright (c) 2025 Charlie Vanaret
# Licensed under the MIT license. See LICENSE file in the project directory for details.

import unopy
Inf = float("inf")

# hs015.mod
def objective(x):
	# the modeler should throw an exception if the function cannot be evaluated
	return 100.*(x[1] - x[0]**2)**2 + (1. - x[0])**2
	
def constraints(x, constraint_values):
	# the modeler should throw an exception if the function cannot be evaluated
	constraint_values[:] = [x[0]*x[1], x[0] + x[1]**2]
	
def objective_gradient(x, gradient):
	# the modeler should throw an exception if the function cannot be evaluated
	gradient[:] = [400.*x[0]**3 - 400.*x[0]*x[1] + 2.*x[0] - 2.,
				   200.*(x[1] - x[0]**2)]

def jacobian(x, jacobian_values):
	# the modeler should throw an exception if the function cannot be evaluated
	jacobian_values[:] = [x[1], 1., x[0], 2.*x[1]]

def lagrangian_hessian(x, objective_multiplier, multipliers, hessian_values):
	# the modeler should throw an exception if the function cannot be evaluated
	hessian_values[:] = [objective_multiplier*(1200*x[0]**2 - 400.*x[1] + 2.),
						-400.*objective_multiplier*x[0] - multipliers[0],
						200.*objective_multiplier - 2.*multipliers[1]]

def lagrangian_hessian_operator(x, evaluate_at_x, objective_multiplier, multipliers, vector, result):
	# the modeler should throw an exception if the function cannot be evaluated
	hessian00 = objective_multiplier*(1200*x[0]**2. - 400.*x[1] + 2.)
	hessian10 = -400.*objective_multiplier*x[0] - multipliers[0]
	hessian11 = 200.*objective_multiplier - 2.*multipliers[1]
	result[:] = [hessian00*vector[0] + hessian10*vector[1],
				hessian10*vector[0] + hessian11*vector[1]]

if __name__ == '__main__':
	# model creation
	problem_type = unopy.PROBLEM_NONLINEAR
	base_indexing = unopy.ZERO_BASED_INDEXING
	# variables
	number_variables = 2
	variables_lower_bounds = [-Inf, -Inf]
	variables_upper_bounds = [0.5, Inf]
	# objective
	optimization_sense = unopy.MINIMIZE
	# constraints
	number_constraints = 2
	constraints_lower_bounds = [1., 0.]
	constraints_upper_bounds = [Inf, Inf]
	number_jacobian_nonzeros = 4
	jacobian_row_indices = [0, 1, 0, 1]
	jacobian_column_indices = [0, 0, 1, 1]
	# Hessian
	number_hessian_nonzeros = 3
	hessian_triangular_part = unopy.LOWER_TRIANGLE
	hessian_row_indices = [0, 1, 1]
	hessian_column_indices = [0, 0, 1]
	lagrangian_sign_convention = unopy.MULTIPLIER_NEGATIVE
	# initial point
	x0 = [-2., 1.]

	model = unopy.Model(problem_type, number_variables, base_indexing)
	model.set_variables_lower_bounds(variables_lower_bounds)
	model.set_variables_upper_bounds(variables_upper_bounds)
	model.set_objective(optimization_sense, objective, objective_gradient)
	model.set_constraints(number_constraints, constraints, constraints_lower_bounds, constraints_upper_bounds,
	  number_jacobian_nonzeros, jacobian_row_indices, jacobian_column_indices, jacobian)
	model.set_initial_primal_iterate(x0)

	# solver creation
	uno_solver = unopy.UnoSolver()
	uno_solver.set_preset("filtersqp")
	uno_solver.set_option("QP_solver", "BQPD")
	print("Solving with Uno", unopy.current_uno_version())

	# run 1: solve with the filtersqp preset with no exact Hessian. Uno defaults to L-BFGS Hessian for NLPs
	result = uno_solver.optimize(model)
	print("Objective at solution:", result.solution_objective)

	# run 2: solve with the filtersqp preset with exact Hessian
	model.set_lagrangian_hessian(number_hessian_nonzeros, hessian_triangular_part, hessian_row_indices,
	  hessian_column_indices, lagrangian_hessian)
	model.set_lagrangian_sign_convention(lagrangian_sign_convention)
	# the Hessian model was overwritten. Set it again
	uno_solver.set_option("hessian_model", "exact")
	result = uno_solver.optimize(model)
	print("Objective at solution:", result.solution_objective)
	# optimization summary
	print("Optimization status:", result.optimization_status)
	print("Solution status:", result.solution_status)
	print("Primal feasibility at solution:", result.solution_primal_feasibility)
	print("Stationarity at solution:", result.solution_stationarity)
	print("Complementarity at solution:", result.solution_complementarity)
	print("Primal solution:", result.primal_solution)
	print("Constraint dual solution:", result.constraint_dual_solution)
	print("Lower bound dual solution:", result.lower_bound_dual_solution)
	print("Upper bound dual solution:", result.upper_bound_dual_solution)
	print("Number of iterations:", result.number_iterations)
	print("CPU time:", result.cpu_time)
	print("Number of objective evaluations:", result.number_objective_evaluations)
	print("Number of constraint evaluations:", result.number_constraint_evaluations)
	print("Number of objective gradient evaluations:", result.number_objective_gradient_evaluations)
	print("Number of Jacobian evaluations:", result.number_jacobian_evaluations)
	print("Number of Hessian evaluations:", result.number_hessian_evaluations)
	print("Number of subproblems solved:", result.number_subproblems_solved)
	assert abs(result.solution_objective - 306.5) <= 1e-4

	# run 3: solve with the ipopt preset
	uno_solver.set_preset("ipopt")
	uno_solver.set_option("linear_solver", "MUMPS")
	result = uno_solver.optimize(model)
	assert abs(result.solution_objective - 306.5) <= 1e-4
	
	# run 4: solve with the filterslp preset
	uno_solver.set_preset("filterslp")
	uno_solver.set_option("LP_solver", "HiGHS")
	result = uno_solver.optimize(model)
	assert abs(result.solution_objective - 306.5) <= 1e-4